tempesta_fw/str.c

```diff
@@ -80,61 +84,153 @@ validate_key(const char *key, int len)
 	validate_cstr(key, len);
 }
 
-#endif /* ifndef DEBUG */
-
-/**
- * Add compound piece to @str and return pointer to the piece.
- */
-TfwStr *
-tfw_str_add_compound(TfwPool *pool, TfwStr *str)
+int
+tfw_str_add_chunk(TfwStr *str, const char *start_pos, const char *end_pos,
+		  TfwPool *pool)
 {
-	validate_tfw_str(str);
-
-	if (unlikely(str->flags & TFW_STR_COMPOUND)) {
-		unsigned int l = str->len * sizeof(TfwStr);
-		unsigned char *p = tfw_pool_realloc(pool, str->ptr, l,
-						    l + sizeof(TfwStr));
-		if (!p)
-			return NULL;
-		str->ptr = p;
-		str->len++;
+	TfwStrChunk *new_chunk, *new_chunks_tbl;
+
+	IF_DEBUG {
+		validate_tfw_str(str);
+		BUG_ON(!start_pos);
+		BUG_ON(!end_pos);
+		BUG_ON(!pool);
+		BUG_ON(str->cnum == TFW_STR_CNUM_MAX);
+		BUG_ON((end_pos - start_pos) > TFW_STR_CHUNK_LEN_MAX);
+		BUG_ON(((end_pos - start_pos) + str->len ) > TFW_STR_LEN_MAX);
+		BUG_ON(end_pos < start_pos);
```

Comment:
BUG_ON() is no-op statemen w/o DEBUG compile option (I mean kernel's, not Tempesta), so there is no need to place BUG_ON() statements into IF_DEBUG. validate_tfw_str() also must be empy statement w/o DEBUG compilation flag.